### PR TITLE
OPENEUROPA-1769: Edit form always moves revision to draft

### PR DESCRIPTION
**Pull request by nagyad**

### CODE REVIEW

**Comment by nagyad** [comment]
After discussing with Tamas the requirement is this: 
Author role is the only role with Edit permission. Edition at any state, always force the content to Draft state.
Reviewer and Validator roles used alone are only high level validation roles to Reject or Approve by moving up the state chain. These roles can edit content only if they are combined with Author role.

I'm implementing the changes accordingly.

File: tests/Behat/FeatureContext.php [hunk 1]
```diff
@@ -68,12 +68,12 @@ public function assertSelectOptions(string $select, TableNode $options): void {
    * @Then the current workflow state should be :state
    */
   public function assertCurrentWorkflowState(string $state): void {
-    $element = $this->getSession()->getPage()->find('css', 'div#edit-moderation-state-0-current');
+    $element = $this->getSession()->getPage()->find('css', 'div#edit-current');
```
Comment (brummbar) [inline tests/Behat/FeatureContext.php]: This is a bit too generic. The id used is quite dynamic: if any other button in the page is rendered before this one, the id will be different. Let's be more precise so we can be future-proof:
```php
public function assertCurrentWorkflowState(string $state): void {
  // Find the content moderation form.
  $xpath = '//form[@class and contains(concat(" ", normalize-space(@class), " "), " content-moderation-entity-moderation-form ")]'
    // Target the text after the "Moderation state" label.
    . '//label[text()="Moderation state"]/following-sibling::text()[1]';
  $element = $this->getSession()->getPage()->find('xpath', $xpath);
  if (empty($element)) {
    throw new \Exception('The current workflow state field is not present on the page.');
  }
  Assert::assertEquals($state, trim($element->getText()));
}
```

File: tests/Behat/FeatureContext.php [hunk 2]
```diff
@@ -68,12 +68,12 @@ public function assertSelectOptions(string $select, TableNode $options): void {
...
     if (empty($element)) {
       throw new \Exception('The current workflow state field is not present on the page.');
     }
     // Need to remove the label from the string.
-    Assert::assertEquals($state, ltrim(trim($element->getText()), 'Current state'));
+    Assert::assertEquals($state, ltrim(trim($element->getText()), 'Moderation state'));
```
Comment (brummbar) [inline tests/Behat/FeatureContext.php]: This is an unsafe usage of `ltrim()`, the second parameter is the character list to remove. In this case it works, but it wouldn't in other cases:
```
>>> ltrim('Moderation state Modified', 'Moderation state');
=> "fied"
```
Anyway with the remark above this part is dropped.

File: tests/features/editorial_workflow.feature [hunk 3]
```diff
@@ -7,71 +7,66 @@ Feature: Corporate editorial workflow
...
+    And I press "Apply"
+    Then I should see the link "Edit"
+    When I click "Edit"
+    Then I should have the following options for the "Change to" select:
+      | Draft |
 
-  Scenario: As a Reviewer user, I can edit demo content in Needs Review state and
-  I can send it back to Draft or to Request Validation
-    Given I am logged in as a user with the "Reviewer" role
+  Scenario: As a Reviewer user, I can moderate demo content by send it back to Draft or to Request Validation
```
Comment (brummbar) [inline tests/features/editorial_workflow.feature]: End the Scenario line with a period.

File: modules/oe_editorial_corporate_workflow/config/install/workflows.workflow.oe_corporate_workflow.yml [hunk 4]
```diff
@@ -63,23 +63,17 @@ type_settings:
...
-      weight: -7
+      weight: -5
     needs_review_to_request_validation:
       label: 'Needs Review to Request Validation'
       from:
         - needs_review
       to: request_validation
-      weight: -5
+      weight: -4
     publish:
```
Comment (brummbar) [inline modules/oe_editorial_corporate_workflow/config/install/workflows.workflow.oe_corporate_workflow.yml]: All the transitions are using the pattern `<from_state>_to_<to_state>` except this one. Let's take the chance to normalise this maybe?

File: modules/oe_editorial_corporate_workflow/oe_editorial_corporate_workflow.module [hunk 5]
```diff
@@ -28,3 +29,16 @@ function oe_editorial_corporate_workflow_menu_local_tasks_alter(&$data): void {
...
   }
 }
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ *
+ * Force the Draft state on node edit forms.
+ */
+function oe_editorial_corporate_workflow_form_node_form_alter(&$form, FormStateInterface $form_state, $form_id) {
+  if (isset($form['moderation_state']['widget'][0]['state']['#options']['draft'])) {
```
Comment (brummbar) [inline modules/oe_editorial_corporate_workflow/oe_editorial_corporate_workflow.module]: With this code, and given the nature of the moderation form that disappears when the node is published, it becomes not possible anymore to put a node in the states "Archived" and "Expired".
This outlines the fact that we miss test coverage for the transitions in those states, we should introduce now these tests.

Comment (nagyad) [reply modules/oe_editorial_corporate_workflow/oe_editorial_corporate_workflow.module]: We have not received clear requirements for those states. We don't know yet what those state will do and it's under design afaik.

File: tests/features/editorial_workflow.feature [hunk 6]
```diff
@@ -7,71 +7,66 @@ Feature: Corporate editorial workflow
...
     Then I should have the following options for the "Save as" select:
-      | Draft        |
-      | Needs Review |
+      | Draft |
     When I fill in "Title" with "Workflow demo"
     And I press "Save"
     And I visit "the content administration page"
     # The content is created and it is not published.
     Then I should see the text "not published" in the "Workflow demo" row
-    When I click "Edit"
```
Comment (brummbar) [inline tests/features/editorial_workflow.feature]: With this change we don't check anymore that an author can edit a node that is in draft. Let's add this back and also check that the only state available in the edit form when a content is in draft is "draft".

File: tests/features/editorial_workflow.feature [hunk 7]
```diff
@@ -7,71 +7,66 @@ Feature: Corporate editorial workflow
...
+    Then I should see the link "Edit"
+    When I click "Edit"
+    Then I should have the following options for the "Change to" select:
+      | Draft |
 
-  Scenario: As a Reviewer user, I can edit demo content in Needs Review state and
-  I can send it back to Draft or to Request Validation
-    Given I am logged in as a user with the "Reviewer" role
+  Scenario: As a Reviewer user, I can moderate demo content by send it back to Draft or to Request Validation
+    Given I am logged in as a user with the "Reviewer" roles
```
Comment (brummbar) [inline tests/features/editorial_workflow.feature]:
```suggestion
    Given I am logged in as a user with the "Reviewer" role
```

File: tests/features/editorial_workflow.feature (hunk 6 above)
Comment (nagyad) [reply tests/features/editorial_workflow.feature]: We do.

Comment (brummbar) [reply tests/features/editorial_workflow.feature]: Where? :)

**Pull request merged**